frontend/src/lib/Quiver.ts

```diff
@@ -131,87 +675,130 @@ export class Quiver {
+      // Blank cells include `blank`.
+      const cssClass = ["blank"]
       if (columnIndex > 0) {
-        classNames.push(`level${rowIndex}`)
+        cssClass.push(`level${rowIndex}`)
       }
+
       return {
-        type: "blank",
-        classNames: classNames.join(" "),
+        type: DataFrameCellType.BLANK,
+        cssClass: cssClass.join(" "),
         content: "",
       }
     }
 
     if (isIndexCell) {
       const dataRowIndex = rowIndex - headerRows
 
-      const uuid = this.styler?.uuid
-      const classNames = [
+      const cssId = this.styler?.uuid
+        ? `T_${this.styler.uuid}level${columnIndex}_row${dataRowIndex}`
+        : undefined
+
+      // Index label cells include:
+      // - row_heading
+      // - row<n> where n is the numeric position of the row
+      // - level<k> where k is the level in a MultiIndex
+      const cssClass = [
         `row_heading`,
         `level${columnIndex}`,
         `row${dataRowIndex}`,
-      ]
+      ].join(" ")
+
+      const contentType = this.types.index[columnIndex].name
+      const content = this.index[dataRowIndex][columnIndex]
 
       return {
-        type: "index",
-        id: uuid
-          ? `T_${uuid}level${columnIndex}_row${dataRowIndex}`
-          : undefined,
-        classNames: classNames.join(" "),
-        // Table index is stored as is (in the column format).
-        content: this.index.data[columnIndex][dataRowIndex],
+        type: DataFrameCellType.INDEX,
+        cssId,
+        cssClass,
+        content,
+        contentType,
       }
     }
 
     if (isColumnsCell) {
       const dataColumnIndex = columnIndex - headerColumns
 
-      const classNames = [
+      // Column label cells include:
+      // - col_heading
+      // - col<n> where n is the numeric position of the column
+      // - level<k> where k is the level in a MultiIndex
+      const cssClass = [
         `col_heading`,
         `level${rowIndex}`,
         `col${dataColumnIndex}`,
-      ]
+      ].join(" ")
 
       return {
-        type: "columns",
-        classNames: classNames.join(" "),
+        type: DataFrameCellType.COLUMNS,
+        cssClass,
         content: this.columns[rowIndex][dataColumnIndex],
       }
     }
 
     const dataRowIndex = rowIndex - headerRows
     const dataColumnIndex = columnIndex - headerColumns
 
-    const uuid = this.styler?.uuid
-    const classNames = ["data", `row${dataRowIndex}`, `col${dataColumnIndex}`]
+    const cssId = this.styler?.uuid
+      ? `T_${this.styler.uuid}row${dataRowIndex}_col${dataColumnIndex}`
+      : undefined
 
+    // Data cells include `data`.
+    const cssClass = [
+      "data",
+      `row${dataRowIndex}`,
+      `col${dataColumnIndex}`,
+    ].join(" ")
+
+    const contentType = this.types.data[dataColumnIndex]
     const content = this.styler?.displayValues
       ? this.styler.displayValues.getCell(rowIndex, columnIndex).content
       : this.data[dataRowIndex][dataColumnIndex]
```

Comment:
When data is  styled, it looks like we return only the styled display value. Should we instead always return `this.data[row][col]` in `content`, and have a second optional `displayContent?: string` field that's set when we have styled data? (This may not be necessary - but it seems like it's possibly important for things like column sorting?)

If we _don't_ do this, I think we should note in the docstring that `content` will always be a formatted string, and not the actual data, for styled dataframes. (And that `contentType` won't refer to `content`'s type, in this case.)

Reply:
Even though sorting can be done with adding a method to Quiver, I agree that this approach is much cleaner and more flexible, i.e. it's done. 😃